### PR TITLE
[SPARK-48505][CORE] Simplify the implementation of `Utils#isG1GC`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR simplifies the implementation of `Utils.isG1GC` as follows：

1. Retain `Utils.classForName("com.sun.management.HotSpotDiagnosticMXBean")`, the main purpose of this step is to detect whether the `HotSpotDiagnosticMXBean` class can be initialized.
2. After successful detection, the code is changed to call the API of `HotSpotDiagnosticMXBean` instead of reflection. 
3. If the detection fails, it will throw a `ClassNotFoundException`, `Utils.isG1GC` will return `false` since this is a NonFatal.

### Why are the changes needed?
Simplify the implementation of `Utils.isG1GC`, try to avoid using reflection as much as possible.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -3058,8 +3059,13 @@ private[spark] object Utils
...
     Try {
-      ManagementFactory.getGarbageCollectorMXBeans.asScala
-        .exists(_.getName.contains("G1"))
+      // SPARK-48505: If the initialization probe of `HotSpotDiagnosticMXBean` is successful,
+      // the reflection API can be avoided in subsequent operations.
+      val clazz = Utils.classForName("com.sun.management.HotSpotDiagnosticMXBean")
+      val hotSpotDiagnosticMXBean = ManagementFactory.getPlatformMXBean(clazz)
+        .asInstanceOf[HotSpotDiagnosticMXBean]
+      val useG1GC = hotSpotDiagnosticMXBean.getVMOption("UseG1GC").getValue
+      "true".equals(useG1GC)
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3068]: cc @rednaxelafx FYI

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3069]: Have I understood your suggestion correctly? But it seems like the detection for class initialization is not necessary here as it's already wrapped in a `Try`?

**Comment by LuciferYang** [comment]
friendly ping @dongjoon-hyun @rednaxelafx Can you help take a look if you have time? Thanks ~

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @LuciferYang .

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -3072,15 +3072,14 @@ private[spark] object Utils
    */
   lazy val isG1GC: Boolean = {
     Try {
+      // SPARK-48505: If the initialization probe of `HotSpotDiagnosticMXBean` is successful,
+      // the API of `HotSpotDiagnosticMXBean` can be used directly in subsequent operations,
+      // instead of invoking it through reflection.
       val clazz = Utils.classForName("com.sun.management.HotSpotDiagnosticMXBean")
-        .asInstanceOf[Class[_ <: PlatformManagedObject]]
-      val vmOptionClazz = Utils.classForName("com.sun.management.VMOption")
+      import com.sun.management.HotSpotDiagnosticMXBean
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3079]: Although the code becomes simpler, I'm not sure this is a better recommended style: `import com.sun.*` in our source code.

There is a similar discussion here.
- https://github.com/apache/spark/pull/43503#discussion_r1671300283

**Comment by dongjoon-hyun** [comment]
BTW, I saw the first attempt here and the comment. 
- #46783

IIUC, he said he is okay with the old code.
> I'd say I'm okay with the old reflection-based version for its preciseness.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3079]: Thanks @dongjoon-hyun ,understood, that makes sense. Let's keep things as they are; I will close this pr.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @LuciferYang .

**Pull request closed without merging**